nvflare/edge/assessors/model_update.py

```diff
@@ -53,10 +56,132 @@ def __init__(
 
+    def _check_device_timeout(self, fl_ctx: FLContext) -> bool:
+        """Check if device wait timeout has been exceeded.
+
+        Args:
+            fl_ctx: FL context
+
+        Returns:
+            bool: True if timeout exceeded, False otherwise
+        """
+        if self.device_wait_start_time is None:
+            return False
+
+        if time.time() - self.device_wait_start_time > self.device_wait_timeout:
+            return True
+        return False
+
+    def _log_device_wait_status(self, fl_ctx: FLContext, message_prefix: str = ""):
+        """Log current device wait status with countdown information."""
+        if self.device_wait_start_time is not None:
+            current_time = time.time()
+            elapsed = current_time - self.device_wait_start_time
+
+            # Only log if we haven't logged recently (rate limiting)
+            if not hasattr(self, "_last_status_log_time"):
+                self._last_status_log_time = 0
+
+            # Adaptive logging frequency based on urgency
+            remaining_time = self.device_wait_timeout - elapsed
+            if remaining_time > 0:
+                # Determine logging interval based on remaining time
+                if remaining_time > 60:  # More than 1 minute: log every 30 seconds
+                    log_interval = 30.0
+                elif remaining_time > 30:  # 30 seconds to 1 minute: log every 15 seconds
+                    log_interval = 15.0
+                elif remaining_time > 10:  # 10 to 30 seconds: log every 5 seconds
+                    log_interval = 5.0
+                else:  # Final 10 seconds: log every 2 seconds
+                    log_interval = 2.0
+
+                # Check if enough time has passed since last log
+                if current_time - self._last_status_log_time >= log_interval:
+                    usable_devices = set(self.device_manager.available_devices.keys()) - set(
+                        self.device_manager.used_devices.keys()
+                    )
+                    self.log_info(
+                        fl_ctx,
+                        f"{message_prefix}Device wait status: "
+                        f"Total devices: {len(self.device_manager.available_devices)}, "
+                        f"usable: {len(usable_devices)}, "
+                        f"expected: {self.device_manager.device_selection_size}. "
+                        f"Timeout in {remaining_time:.1f} seconds.",
+                    )
+                    self._last_status_log_time = current_time
+
+    def _start_timeout_tracking(self, fl_ctx: FLContext):
+        """Start independent timeout tracking using thread pool."""
+        if self.timeout_future is not None and not self.timeout_future.done():
+            # Cancel existing timeout if still running
+            self.timeout_future.cancel()
+
+        # Submit new timeout task to thread pool
+        self.timeout_future = self.thread_pool.submit(self._timeout_tracker, fl_ctx)
```

Comment:
this is changed both in main thread and this worker thread.

Reply:
updated